### PR TITLE
Speeding up the system tests further

### LINKED ISSUE #23373: SystemTests read, write and delete data files with the same names
### Expected behavior

`SystemTests` should be writing data in a single directory to facilitate management of the outputs. Preferably, each test should also output data with filenames specific to that test.

### Actual behavior

When trying to run `systemtests` in parallel splitting the list of tests across multiple instances (akin to the `ctest` `-j` option), it was found (see https://github.com/mantidproject/mantid/pull/23286) that different test modules (= `.py` files in the `Testing/SystemTests/tests/analysis` directory) are writing in the same directories, using the same files names.
There are at least 3 different locations the tests write to:
1. `build/ExternalData/Testing/Data/SystemTest`
2. `build/Testing/SystemTests/scripts`
3. `build`

In some cases a process tries to write to a file that already exists (same name used by another test).
In other cases a process performs a `.cleanup()` which removes a file that is still needed by a running test.

The solution was to group the tests in lists where all modules starting with the same 4 letters are given to one core. This worsens the load balance between cores (with 8 cores, core 1 performs 93 tests while core 8 only has 44). This is not ideal but allows the suite to complete without failures.

This seems to occur mostly for `SANS/SANS2D` and `ISIS` tests.
It would be nice to clean up the places where the tests write to, so it could be controlled so that each core is writing to its own directory and performing cleanups will not affect other processes that are running in different directories.

### PR DESCRIPTION
**Pull request by nvaytet**
**Description of work.**

The scheduling of the system tests is now done in the following way.:

A master test list is first built from scanning the test python files in the `Testing/SystemTests/tests/analysis` directory. For each module in this list, a list of data files required by that module is created. An accompanying lock status for each data file is also created.

Then, N processes are spawned, where N is the number requested by the user via the `-j N` option.

Each thread starts a loop and gathers a first test module from the master test list which is stored in the tests_dict shared dictionary, starting with the number in the module list equal to the process id.

Each process then checks if all the data files requird by the current test module are available (i.e. have not been locked by another thread). If all files are unlocked, the thread proceeds with that test module. If not, it goes further down the list until it finds a module whose files are all available.

Once it has completed the work in the current module, it checks if the number of modules that remains to be executed is greater than 0. If there is some work left to do, the thread finds the next module that still has not been executed (searches through the tests_lock array and finds the next element that has a 0 value).

This aims to have all threads end calculation approximately at the same time, and also avoids having to modify many save directories in the test files and the facilities interfaces they use.

**To test:**

`./systemtests -j 8 -q --output-on-failure` should run even faster than before.
I have timed it at ~18min using `-j 12`. The longest test module teakes ~16min to complete so it will not be possible to go below that.

Fixes #23373 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Testing/SystemTests/lib/systemtests/stresstesting.py [hunk 1]
```diff
@@ -71,6 +72,16 @@ def __init__(self):
         from mantid.kernel import MemoryStats
         self.memory = MemoryStats().residentMem()/1024
 
+    def setSearchSaveDirectories(self, core_id, saveDir):
+        '''
+        Modify the defaultsave.directory to a path specific to each thread
+        '''
+        from mantid.kernel import config
+        config['datasearch.directories'] += "/core-%i" % (core_id)
+        config['defaultsave.directory'] = "%s/core-%i" % (saveDir,core_id)
```
Comment (martyngigg) [inline Testing/SystemTests/lib/systemtests/stresstesting.py]: Are we going to need to warn developers now that `config['defaultsave.directory']` is now only fixed at runtime so it shouldn't be cached in the test?

**Comment by nvaytet** [comment]
Hi @martyngigg I've now changed things around again, in a way which is much better in my opinion.

As the description of work says above, I am now creating a list of data files that each test module requires at runtime.

The threads pick the next module in a global test list that has not yet been run and checks if any of the data files required by that module have not been locked by another threads.

If all files are unlocked, it locks them and proceeds with running the tests in that module. If any one of the files is locked, then it moved on to the next module in the list to see if that one can be run. This no longer required any messing around with default save directories etc...

This is of course as long as I have correctly identified what is a data file in the python scripts. For now I am checking for anything that has a '.nxs' extension and any sequence of 4 or 5 digits inside a string (e.g. "54431" or "TOPAZ_3132"). Maybe I also need to add the ".raw" extension?

It works pretty well on my machine. Down to 18min on 12 cores.

**Comment by peterfpeterson** [comment]
FYI #23458 is likely going to get merged before this and resolving the conflicts will be fun.

**Comment by nvaytet** [comment]
@peterfpeterson aaaarrgh! :-)

**Comment by peterfpeterson** [comment]
If you are going to clean up the file stuff [use the `with` keyword](http://effbot.org/zone/python-with-statement.htm) which will close the file even in the case of exceptions.
```python
with open("x.txt") as f:
    data = f.read()
    do something with data
```
It works for writing too.

**Comment by nvaytet** [comment]
I think I used the `with open(...) as f` syntax as much as possible now.

**Comment by peterfpeterson** [comment]
You missed one in L970 of `Testing/SystemTests/lib/systemtests/stresstesting.py`

**Comment by nvaytet** [comment]
Yes, but since you hadn't modified it in your PR, I left it alone. I thought you decided it was better that way (https://github.com/mantidproject/mantid/pull/23458/files#diff-5247c1f456dfe545c21d1a44924ad807R923)

**Comment by nvaytet** [comment]
@peterfpeterson  I fixed them all now

File: Testing/SystemTests/lib/systemtests/stresstesting.py [hunk 2]
```diff
@@ -759,117 +762,159 @@ def __init__(self, test_loc, runner, output=[TextResultReporter()],
...
+                                if not check.search(line[i]):
+                                    key = line[i+1:indx]+ext
+                                    if (key not in files_required_by_test_module[modkey]) and (key != ext):
+                                        files_required_by_test_module[modkey].append(key)
+                                        data_file_lock_status[key] = False
+                                    break
+
+                    # Search for '0123 or "0123
+                    for so in string_quotation_mark:
+                        p = re.compile(so+r"\d\d\d\d")
```
Comment (peterfpeterson) [inline Testing/SystemTests/lib/systemtests/stresstesting.py]: Can be `p = re.compile(so+r"\d{4}")`

File: Testing/SystemTests/lib/systemtests/stresstesting.py [hunk 3]
```diff
@@ -759,117 +762,159 @@ def __init__(self, test_loc, runner, output=[TextResultReporter()],
...
+                                if line[i] == so:
+                                    key = line[m.start()+1:i]
+                                    if key not in files_required_by_test_module[modkey]:
+                                        files_required_by_test_module[modkey].append(key)
+                                        data_file_lock_status[key] = False
+                                    break
+
+                    # Search for 0123' or 0123"
+                    for so in string_quotation_mark:
+                        p = re.compile(r"\d\d\d\d"+so)
```
Comment (peterfpeterson) [inline Testing/SystemTests/lib/systemtests/stresstesting.py]: Can be `p = re.compile(r"\d{4}"+so)`

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Small regular expression change then it can be merged. Good improvement to the code in general. I'm still surprised there isn't a standard python implementation of a thread pool.

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**